autorest/client.go

```diff
@@ -140,12 +141,32 @@ type Client struct {
 // NewClientWithUserAgent returns an instance of a Client with the UserAgent set to the passed
 // string.
 func NewClientWithUserAgent(ua string) Client {
-	return Client{
+	c := Client{
 		PollingDelay:    DefaultPollingDelay,
 		PollingDuration: DefaultPollingDuration,
 		RetryAttempts:   DefaultRetryAttempts,
 		RetryDuration:   30 * time.Second,
-		UserAgent:       ua,
+		UserAgent:       getDefaultUserAgent(),
+	}
+	c.AddToUserAgent(ua)
+	return c
+}
+
+// getDefaultUserAgent builds a string containing the Go version, system archityecture and OS,
+// and the go-autorest version.
+func getDefaultUserAgent() string {
+	return fmt.Sprintf("Go/%s (%s-%s) go-autorest/%s",
+		runtime.Version(),
+		runtime.GOARCH,
+		runtime.GOOS,
+		Version(),
+	)
+}
+
+// AddToUserAgent adds an extension to the current user agent
+func (c *Client) AddToUserAgent(extension string) {
```

Comment:
Personally, I think it may be good to return an error here if `extension` is the empty string.